magma/compile_guard.py

```diff
@@ -0,0 +1,116 @@
+import dataclasses
+from typing import List
+
+from magma.array import Array
+from magma.bits import Bits
+from magma.clock import Clock
+from magma.circuit import Circuit, CircuitBuilder, _DefinitionContextManager
+from magma.digital import Digital
+from magma.ref import AnonRef, ArrayRef, DefnRef, InstRef, TupleRef
+from magma.t import In
+from magma.tuple import Tuple
+
+
+@dataclasses.dataclass(frozen=True)
+class _CompileGuardState:
+    ckt: CircuitBuilder
+    ctx_mgr: _DefinitionContextManager
+
+
+class _CompileGuardBuilder(CircuitBuilder):
+    def __init__(self, name, cond):
+        super().__init__(name)
+        self._cond = cond
+        self._system_types_added = {}
+        self._set_inst_attr("coreir_metadata", {"compile_guard": cond})
+
+    def _rewire(self, port, value):
+        new_port_name = f"__{id(port.name.inst)}_{port.name.name}"
+        T = type(value).undirected_t
+        self._add_port(new_port_name, In(T))
+        new_port = self._port(new_port_name)
+        port.unwire(value)
+        port @= new_port
+        external = getattr(self, new_port_name)
+        external @= value
+
+    def _pre_finalize(self):
+        for inst in self._instances:
+            self._process_instance(inst)
+
+    def _process_instance(self, inst):
+        for port in inst.interface.inputs(include_clocks=True):
+            self._process_port(port)
+
+    def _process_port(self, port, ref=None, kwargs=None):
+        value = port.value()
+        if ref is None:
+            if value is None:
+                self._process_undriven(port)
+                return
+            ref = value.name
+        if isinstance(ref, AnonRef):
+            if value.const():
+                return
+            # TODO(rsetaluri): Implement valid anon. values.
+            raise NotImplementedError()
```

Comment:
One challenging case to think about here is if the anon value is wired up after the compile guard block.  It may come from an external driver but that driver might not be created yet.  I think because we assume that values only flow into the compile guard, if we encounter an undriven anon, we can assume the external driver case, but we should be sure to test this.

Reply:
Yeah that's a good point. So if I understand correctly you're talking about the case like

```python
class Top:
    ...
    value = ...
    tmp = T()
    with compile_guard(...):
        f.assert_immedate(tmp == 0)
    tmp @= value
```

Good point. This could be part of the `compile_guard` spec. If it really turns out that people want to do this frequently, then I think we could make all of this logic a late(r) binding pass as you had it. That would of course be more robust to such things.